### ISSUE TITLE
runtime: frequent SIGSEGV in readgstatus in bootstrap toolchain

### ISSUE THREAD

**Issue by bcmills** [golang/go#49686]
`greplogs --dashboard -md -l -e '(?ms)^fatal error: .*^runtime\.readgstatus' --since=2021-01-01`

<details>

[2021-11-18T22:40:42-2375b6e/linux-mips64-rtrk](https://build.golang.org/log/86e5b25c30b9b986f26c3d2d9003922473893827)
[2021-11-18T19:16:38-c4aae23/linux-mips64le-rtrk](https://build.golang.org/log/eb7822342e94251a6771cfa89cffb475e1002df9)
[2021-11-15T23:01:05-d156101/linux-mips64-rtrk](https://build.golang.org/log/aff7b5b7f22f21a744dd3af18446509393a24b7d)
[2021-11-12T16:58:34-95d0657/linux-mips64le-rtrk](https://build.golang.org/log/971f4ba442a3a4fef7b3be8aaaa4d6b012127195)
[2021-11-11T11:47:33-c49627e/linux-mips64le-rtrk](https://build.golang.org/log/f8df602d0936fd5afe486acaa4f23ee1c112736b)
[2021-11-10T05:08:25-17980df/linux-mips64le-rtrk](https://build.golang.org/log/2794bd4a18937b6e854fdeb66686c72457ec5012)
[2021-11-09T22:11:33-4aa0746/linux-arm64-packet](https://build.golang.org/log/d3983e9f11d4bafa723a9025f884c6ea9245b1ba)
[2021-11-09T16:20:19-e90fd9a/linux-arm64-packet](https://build.golang.org/log/e6b4557c23d696c38177c110e6b6cdfee5da2105)
[2021-11-04T20:43:07-8248152/linux-arm64-packet](https://build.golang.org/log/8b270cc799e86e25abfb72053c970900c097e39e)
[2021-11-02T16:12:23-599de4b/linux-arm64-packet](https://build.golang.org/log/5a5e2ec6e6817b1ce4d9bd5ad3001610ef4b45a3)
[2021-11-01T21:27:46-036812b-f9cb33c/linux-arm64-packet](https://build.golang.org/log/edcfe6ba3c8bf2dac4aee02ad48fa4f3d3fdac53)
[2021-10-28T16:54:29-5fce1d9/linux-arm64-packet](https://build.golang.org/log/309bec6b1b971d8de475e1e0a2a4931ed795a46e)
[2021-10-26T05:02:53-ec6c004/linux-arm64-packet](https://build.golang.org/log/de5d36407a4693b1e3879e1d12b6f0e29d52ad8b)
[2021-10-15T07:13:43-1cbec68/linux-arm64-packet](https://build.golang.org/log/b5ce3f8ec0655129104e7142acd063d5a9ed62fd)
[2021-10-11T22:32:23-c1b0ae4/linux-arm64-packet](https://build.golang.org/log/caf40e8ce4d8234d66bd0ab0806f8a774216016c)
[2021-10-08T17:58:41-59d4e92-16a3cef/android-amd64-emu](https://build.golang.org/log/aa2cb0364ec64c5f0a2bf61e38129c5776d306c7)
[2021-10-08T00:24:52-59d4e92-78d749f/android-amd64-emu](https://build.golang.org/log/33fe69416614480c2461d69ae33157b26adf2e2f)
[2021-10-08T00:18:29-59d4e92-7cef831/android-amd64-emu](https://build.golang.org/log/e20ff16e03cd274b95ea37f260cae106abe424f3)
[2021-10-07T21:39:17-59d4e92-6b4cf2b/darwin-arm64-11_0-toothrot](https://build.golang.org/log/e3143695cab88888688708c9c2e9c0959310ddd4)
[2021-10-07T18:58:33-7286502/darwin-arm64-11_0-toothrot](https://build.golang.org/log/8295194cfe70a64d8649098f65a836ad41f5e7fd)
[2021-10-07T18:58:33-7286502/linux-arm64-packet](https://build.golang.org/log/cddca82008550eaf192503d0627860c93bb77364)
[2021-10-07T18:58:33-59d4e92-7286502/android-amd64-emu](https://build.golang.org/log/cf44727199f060e3e8f138970ee9abd3599e8f25)
[2021-10-07T18:45:53-77f2750/android-arm64-corellium](https://build.golang.org/log/405a318eb555560810fb5e4a78689a5b72897a87)
[2021-10-07T18:45:53-59d4e92-77f2750/android-amd64-emu](https://build.golang.org/log/6b708fa45221808e9393093816f0b75f74b79650)
[2021-10-07T14:01:52-ecb2f23/android-386-emu](https://build.golang.org/log/7b1bb5ae0623e5f35ee2dd36e4066ae51ee65792)
[2021-10-07T03:41:40-62292e8-6f74ed0/android-amd64-emu](https://build.golang.org/log/2187589a735d2d8e539216f3f070b74d5c639537)
[2021-10-07T02:30:47-62292e8-39bbf08/darwin-arm64-11_0-toothrot](https://build.golang.org/log/994761ea6024034f8375dce791645bb5570f907f)
[2021-10-07T00:39:17-62292e8-812a33d/aix-ppc64](https://build.golang.org/log/a646fbe9017a00a63228fe7c8e2e9105a25a0d4c)
[2021-10-06T23:01:01-62292e8-4ffa2f1/android-amd64-emu](https://build.golang.org/log/f68d93380c4194fb82f22bb2c01d1b1ca6a517cc)
[2021-10-06T22:42:28-62292e8-4002616/darwin-arm64-11_0-toothrot](https://build.golang.org/log/494da6b4ed4730c5a76ca7c678ded7f596b0cdd4)
[2021-10-06T21:17:08-62292e8-2e107b4/android-amd64-emu](https://build.golang.org/log/c443a4442d00c324be6e09f70d5e3bc401493531)
[2021-10-06T20:30:12-62292e8-4a37a1d/android-amd64-emu](https://build.golang.org/log/1cf87f8ecf85a9abbab51b337add94b79f052aa0)
[2021-10-06T19:59:27-62292e8-17c513e/android-amd64-emu](https://build.golang.org/log/e74bae9e645d36ebdfca0d70b2d71a492c3f2481)
[2021-10-06T19:59:27-62292e8-17c513e/darwin-arm64-11_0-toothrot](https://build.golang.org/log/0f8f5249f25da0508ae589727d418a7fadb3a297)
[2021-10-06T17:38:01-d2e5035-415f0a3/android-amd64-emu](https://build.golang.org/log/089db99cfd4d093aed6bd59ef025d66a350ee906)
[2021-10-06T16:46:09-d2e5035-3160571/android-amd64-emu](https://build.golang.org/log/621e624cce30ef8c0a6158bf1710fe2e32984e03)
[2021-10-05T23:58:29-d2e5035-ac60900/android-amd64-emu](https://build.golang.org/log/f96c8dd0a6cc9d6fc63cd893eb31a4855a707dc1)
[2021-10-05T21:25:51-6ae3afa/darwin-arm64-11_0-toothrot](https://build.golang.org/log/e4f355784329dc7d9b3a061aaa538d7d9bd3893c)
[2021-10-05T20:53:02-d4b1ae0-695a59b/android-amd64-emu](https://build.golang.org/log/d47b97e5535f7b26bfc0082242cc3795760efaa4)
[2021-10-05T20:35:54-d4b1ae0-990c9c6/darwin-arm64-11_0-toothrot](https://build.golang.org/log/b438cb54a33b9b2bcf407bf6a44a65e0ac9befa8)
[2021-10-05T20:15:01-d4b1ae0-0b4d499/android-amd64-emu](https://build.golang.org/log/2872debf7408d9df2b0d2144c299132592ba7dc6)
[2021-10-05T20:15:01-d4b1ae0-0b4d499/darwin-arm64-11_0-toothrot](https://build.golang.org/log/496dbb4450c320a3a0fd1cb6018ae295fd6a36b8)
[2021-10-05T19:28:36-d4b1ae0-7ae83c8/android-amd64-emu](https://build.golang.org/log/d70e84921281e2a142ad5b30226315c575fa3796)
[2021-09-12T01:06:53-915f620-0d8a4bf/linux-arm64-packet](https://build.golang.org/log/93d10742f91534cfe9c0fc38f667ea86caedd065)
[2021-08-31T08:41:16-3e0d083-f118d14/linux-arm64-packet](https://build.golang.org/log/176928e36317079117a6f00d9e2ba91153fb1c4e)
[2021-08-08T18:37:08-507cc34/linux-mips64-rtrk](https://build.golang.org/log/62482472766ced7a1a6453ae2812879ad95b4d6b)
[2021-06-25T17:31:39-d01bc57/linux-mips64-rtrk](https://build.golang.org/log/0bbb1c068d3f04af834235e3fd01da1a394b7c6e)
[2021-06-07T21:12:46-909dd5e/linux-arm64-packet](https://build.golang.org/log/1a339d23466bdf343bfd145a47e5b110c2ab89de)
[2021-05-26T16:11:00-bfd7798/linux-mips64le-rtrk](https://build.golang.org/log/c1de6acf5cccb51ebbc1e9a3aa9548a674f6d0ea)
[2021-05-21T21:07:03-3c65644/linux-mips64-rtrk](https://build.golang.org/log/8bf7193b69d7d5886d075c784d8df70df28c59bf)
[2021-05-17T18:03:56-a2c07a9/linux-mips64-rtrk](https://build.golang.org/log/648d39a8eac0f222329134d2109d09427a2d8920)
[2021-05-07T18:14:25-0714010-af6123a/linux-mips64-rtrk](https://build.golang.org/log/afc1a10ecd07626e424baec56db16498851c20d0)
[2021-05-07T17:43:31-b44c78b/linux-arm64-packet](https://build.golang.org/log/dda98b06ae3c735c4d3cc8a2bc875d20e946a219)
[2021-04-23T20:57:54-691e1b8/linux-mips64le-rtrk](https://build.golang.org/log/545943f9836776fe9fc8a7ab1b64e1de690c4556)
[2021-04-16T21:20:31-02a2ff4/linux-mips64-rtrk](https://build.golang.org/log/f36b10a56347d13647cdfeffc072a86491295039)
[2021-03-25T13:57:08-d1beb07-9f4d5c9/linux-mips64-rtrk](https://build.golang.org/log/5c67738fbe440dff8a803cf2aecdfa432022a0c0)
[2021-03-17T00:17:33-77fc1ea-68f8e1a/linux-mips64le-rtrk](https://build.golang.org/log/ad1130d052717336f16e7818ca6d2b24c95c8f02)
[2021-03-12T01:47:01-a607408/linux-mips64le-rtrk](https://build.golang.org/log/8340761e4e4d874bff80180b8c405cd1a46912f6)
[2021-03-08T20:41:06-a08adda/linux-mips64le-rtrk](https://build.golang.org/log/3218c53caac1a1699a9267f1e8fa97d174805d7f)
[2021-03-05T18:46:36-51d8d35/linux-mips64le-rtrk](https://build.golang.org/log/0bb9a1381aa9c98431f2b228a4e2ffb70995146d)
[2021-02-23T10:04:13-a78b0e6/linux-mips64-rtrk](https://build.golang.org/log/f22c48a4a5864215644459b7536649a8dc33c378)
[2021-01-27T12:18:00-cd176b3/linux-mips64-rtrk](https://build.golang.org/log/dc680de7f914029f2b6226800756ddd5c6b179d0)
[2021-01-19T12:49:13-d047c91/linux-arm64-packet](https://build.golang.org/log/04aed0f72cd1e842191c17c0b7af4245f7a226f3)

</details>

Note that several of these failures are on `linux/arm64` and `darwin/arm64`, which are [first-class ports](https://golang.org/wiki/PortingPolicy#first-class-ports).

(CC @cherrymui @jeremyfaller)

**Comment by jeremyfaller** [comment]
Also, CC – @mknyszek

**Comment by cherrymui** [comment]
I'll take a look.

**Comment by cherrymui** [comment]
This looks like #43824 . The most recent ones are all failing in markroot with a nil gp, which is the same as #43824.

**Comment by cherrymui** [comment]
> why is the failing address exactly zero? Shouldn't the address be slightly offset for the atomicstatus field? Is there a nil pointer check here, before the load?

I looked into this a while ago. Yes, there is a nil check. At least on some architectures, atomic load instruction is not marked "fault on nil" in the compiler, so the compiler does not eliminate the nil check. We probably want to fix that (but that won't fix the issue).

**Comment by aclements** [comment]
I can definitely see this failing on a weak memory machine. I don't think it even requires subtlety around STW.

1. We start a GC. `gcMarkRootPrepare` snapshots allglen atomically. We restart the world.
2. Thread 1 does `allgadd`, which appends to `allgs`. Suppose this requires a grow, so thread 1 allocates a new all-nil slice, copies the old slice into the new slice (write A), and assigns the new slice to allgs (write B).
3. Thread 2 concurrently enters `markroot` and reads from allgs. It reads the slice base (read A) and then the n'th value in the slice (read B).

On a weak memory machine, thread 2 could observe write B before write A (or, equivalently, the reads could be reordered), leading to thread 2 reading a nil from `allgs`.

**Comment by aclements** [comment]
I've got a couple ideas for how to fix this. Probably the most straightforward is to save a copy of allgs during gcMarkRootPrepare and use that in markroot.

**Comment by cherrymui** [comment]
Does it work if we use `atomicAllGIndex` ? `allgptr` is written atomically so if it points to the new memory it _should_ observe the writes to the array?

**Comment by aclements** [comment]
Probably, though then we also have to call `atomicAllG` either on every markroot or save that somewhere. (In general I really don't like the atomicAllG API... I might be happier with that if atomicAllG just returned a slice.)

**Comment by mknyszek** [comment]
[2022-09-05T08:28:34-83b083e-bd5595d/linux-mips-rtrk](https://build.golang.org/log/22119320d8db9581db841b19982813963085f1a6)

This looks like it may be related? It's a nil pointer dereference in `readgstatus` from `markroot`.

**Comment by cherrymui** [comment]
It looks like cmd/dist crashed, which is built from the bootstrap toolchain, which is go1.17.13, which doesn't seem to include the fix above.

**Comment by bcmills** [comment]
Given that it affects the bootstrap toolchain, should the fix be backported to `release-branch.go1.17`? (Or would we rather live with the crashes until the bootstrap version is raised again?)

**Comment by randall77** [comment]
I think we should backport the fix.

Normally we just add a workaround to tip's toolchain to avoid bootstrap bugs, but I suspect there isn't a feasible workaround in this case.

**Comment by cherrymui** [comment]
Go 1.17 is no longer supported. Not sure how we're going to backport. I guess we could cherry-pick to release-branch.go1.17. We still need to manually update the bootstrap toolchains on the builders, right?

**Comment by randall77** [comment]
We can release another 1.17.x even though it is not supported. Didn't we do that for 1.4.x? Or maybe I am misremembering.

**Comment by rsc** [comment]
We never did an updated 1.4.x release. We only backported things to the release branch and posted new src tar files. 

That said, we're in a brave new world and we can chart whatever course we think best. Since this seems to be entirely the relaxed memory model architectures, one option is to bump all those builders to Go 1.18 or Go 1.19. We'll still have plenty on Go 1.17 to keep us honest (a few builders are already bumped for other reasons, like the port being new since Go 1.17).

If people generally agree with bumping the bootstrap toolchain forward on those systems, I'm happy to do the work or advise someone who wants to learn the process.

**Comment by mknyszek** [comment]
Sounds like we have 3 paths forward here:
1. Do nothing, it's not that common and we may want to update the bootstrap toolchain in the next release anyway.
2. Backport the fix to the bootstrap compiler.
3. Move just the relaxed memory architecture bootstrap toolchains up a version or two.

I think this just needs a decision and someone to do the work (or no one to do no work).

**Comment by rsc** [comment]
I talked with Austin briefly about this. I think (2) is probably wrong; it's strictly more work than (3). 
I don't think we should do (1) and leave lingering flakes (assuming flakes are still happening), so I'm inclined to (3).

**Comment by rsc** [comment]
Re "assuming flakes are still happening", taking a closer look at the history, it looks like there were many failures during run.bash originally, and that got fixed and released. 

Now we have a single flake where readgstatus faulted in readgstatus in a similar way, during bootstrap, which suggests maybe older toolchains are problematic. But it's just one flake. It might be best to wait for a couple more to really be sure that this is the problem. I did a greplogs for runtime.readgstatus, and I found this fault too, which is from a commit back in May (not too old) and was during run.bash, so not bootstrap-dependent. It's interesting because it is pprof faulting on readgstatus, while the world is ostensibly stopped. 

[2022-05-09T20:57:58-3f43096/openbsd-arm-jsing](https://build.golang.org/log/58f8d59bcd5d46aaea7daaad80b48f16a872129f)

Maybe there is something lurking here, or maybe not. Let's keep accumulating failures mentioning runtime.readgstatus and see what we find.
